sources/whippet/src/main.rs

```diff
@@ -0,0 +1,164 @@
+/*!
+whippet is a D-Bus listener that reponds to events on D-Bus.
+Currently, it is used to notify netdog of events for the primary interface on the network1 D-Bus service.
+*/
+use futures::stream::StreamExt;
+use snafu::{ensure, ResultExt};
+use std::process;
+use std::process::Command;
+use std::string::String;
+use zbus::zvariant::OwnedObjectPath;
+use zbus::Connection;
+
+mod networkmanager;
+
+use networkmanager::{LinkState, NetworkManagerProxy};
+
+static NETDOG: &str = "/usr/bin/netdog";
+
+pub(crate) async fn run() -> Result<(), crate::error::Error> {
+    // Connect to the system bus
+    let conn = Connection::system().await?;
+
+    // Get the manager created from the connection to the system bus
+    let manager = NetworkManagerProxy::new(&conn).await?;
+
+    // List the lists on the bus
+    let links = manager.list_links().await?;
+
+    // Call netdog primary-interface to get the name of the primary interface
+    println!("Calling netdog to get primary interface");
+    let primary_interface_name_result = Command::new(NETDOG)
+        .arg("primary-interface")
+        .output()
+        .context(error::NetdogExecutionSnafu)?;
+    ensure!(
+        primary_interface_name_result.status.success(),
+        error::FailedNetdogSnafu {
+            stderr: String::from_utf8_lossy(&primary_interface_name_result.stderr)
+        }
+    );
+
+    let primary_interface_output_str = String::from_utf8(primary_interface_name_result.stdout)
+        .context(error::PrimaryInterfaceStringSnafu {})?;
+    let primary_interface: String = primary_interface_output_str
+        .trim()
+        .to_lowercase()
+        .trim_matches('"')
+        .to_string();
+    println!("Primary interface is {}", &primary_interface);
+
+    // Put the path in an option since we might not find it
+    let mut path_to_primary: Option<&OwnedObjectPath> = None;
+
+    // Iterate over the links found by list_links()
+    for (id, name, path) in links.iter() {
+        println!("Link id: {id} Name: {name} Path: {path:?}");
+        if name == &primary_interface {
+            // Now grab the Properties path from the listing to start watching for changes
+            path_to_primary = Some(path);
+            let link_status: LinkState =
+                serde_json::from_str(manager.describe_link(*id).await?.as_str())
+                    .context(error::DbusDescribeLinkSnafu {})?;
+            println!(
+                "Found {} is {}",
+                &primary_interface, link_status.administrative_state
+            );
+            if link_status.administrative_state == "configured" {
+                // call netdog now since its already configured, then the polling for changes can block
+                println!(
+                    "Calling netdog write-primary-interface-status for {}",
+                    link_status.name
+                );
```

Comment:
This is racy because the interface may not be configured here, but may become configured before we are listening for the signal that tells us it's now configured.